File: problems/tier2-intermediate/017-lru-cache/solution.cpp
#include <iostream>
#include <memory>
#include <unordered_map>
#include <vector>
#include <algorithm>
using namespace std;

// ─── Data Model ─────────────────────────────────────────────────────────────

struct Node {
    int key;
    int value;
    long expiresAt;
    Node* prev;
    Node* next;
    Node(int k, int v, long exp = 0)
        : key(k), value(v), expiresAt(exp), prev(nullptr), next(nullptr) {}
};

enum class EvictionReason { CAPACITY, TTL_EXPIRED, EXPLICIT_DELETE };

// ─── Eviction Listener Interface ─────────────────────────────────────────────

class EvictionListener {
public:
    virtual void onEviction(int key, int value, EvictionReason reason) = 0;
    virtual ~EvictionListener() = default;
};

// ─── LRU Cache ──────────────────────────────────────────────────────────────

class LRUCache {
private:
    int capacity;
    unordered_map<int, Node*> cache;
    Node* head;
    Node* tail;
    vector<EvictionListener*> listeners;
    bool notifying = false;

    void addToFront(Node* node) {
        node->next = head->next;
        node->prev = head;
        head->next->prev = node;
        head->next = node;
    }

    void removeNode(Node* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }

    void moveToFront(Node* node) {
        removeNode(node);
        addToFront(node);
    }

    bool isExpired(Node* node, long currentTime) {
        return node->expiresAt > 0 && currentTime >= node->expiresAt;
    }

    void notifyListeners(int key, int value, EvictionReason reason) {
        notifying = true;
        for (auto* listener : listeners) {
            listener->onEviction(key, value, reason);
        }
        notifying = false;
    }

    void evictNode(Node* node, EvictionReason reason) {
        removeNode(node);
        cache.erase(node->key);
        notifyListeners(node->key, node->value, reason);
        delete node;
    }

    void evictLRU() {
        if (head->next == tail) return;  // empty
        Node* lru = tail->prev;
        evictNode(lru, EvictionReason::CAPACITY);
    }

public:
    LRUCache(int cap) : capacity(cap) {
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head->next = tail;
        tail->prev = head;
    }

    ~LRUCache() {
        Node* curr = head;
        while (curr) {
            Node* next = curr->next;
            delete curr;
            curr = next;
        }
    }

    // Part 1 interface
    int get(int key) {
        if (cache.find(key) == cache.end()) return -1;
        Node* node = cache[key];
        moveToFront(node);
        return node->value;
    }

    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            Node* node = cache[key];
            node->value = value;
            moveToFront(node);
            return;
        }
        if ((int)cache.size() >= capacity) {
            evictLRU();
        }
        Node* newNode = new Node(key, value);
        addToFront(newNode);
        cache[key] = newNode;
    }

    // Part 2 interface
    int get(int key, long currentTime) {
        if (cache.find(key) == cache.end()) return -1;
        Node* node = cache[key];
        if (isExpired(node, currentTime)) {
            evictNode(node, EvictionReason::TTL_EXPIRED);
            return -1;
        }
        moveToFront(node);
        return node->value;
    }

    void put(int key, int value, long currentTime, int ttl = 0) {
        long expiresAt = (ttl > 0) ? currentTime + ttl : 0;

        if (cache.find(key) != cache.end()) {
            Node* node = cache[key];
            if (isExpired(node, currentTime)) {
                evictNode(node, EvictionReason::TTL_EXPIRED);
            } else {
                node->value = value;
                node->expiresAt = expiresAt;
                moveToFront(node);
                return;
            }
        }
        if ((int)cache.size() >= capacity) {
            evictLRU();
        }
        Node* newNode = new Node(key, value, expiresAt);
        addToFront(newNode);
        cache[key] = newNode;
    }

    bool deleteKey(int key) {
        if (cache.find(key) == cache.end()) return false;
        evictNode(cache[key], EvictionReason::EXPLICIT_DELETE);
        return true;
    }

    int size() {
        return (int)cache.size();
    }

    // Part 3 interface
    void addEvictionListener(EvictionListener* listener) {
        listeners.push_back(listener);
    }

    void removeEvictionListener(EvictionListener* listener) {
        listeners.erase(
            remove(listeners.begin(), listeners.end(), listener),
            listeners.end());
    }
};

// ─── Ops simulator (used by spec-based tests) ───────────────────────────────
//
// LRUCache is a stateful "design" problem. The simulator drives a single
// LRUCache instance through a list of operations and returns one string per
// operation describing its outcome.
//
// Op fields (ints in i1..i4, string in s1):
//   "new"            — create cache(capacity=i1)              -> "ok"
//   "put"            — put(i1, i2)                            -> "ok"
//   "put_t"          — put(i1, i2, i3=now, i4=ttl)            -> "ok"
//   "get"            — get(i1)                                -> int as string
//   "get_t"          — get(i1, i2=now)                        -> int as string
//   "delete"         — deleteKey(i1)                          -> "ok"/"fail"
//   "size"           —                                         -> int as string
//   "add_listener"   — i1 = listener slot index                -> "ok"
//   "remove_listener"— i1 = listener slot index                -> "ok"
//   "events_count"   — i1 = listener slot index                -> int as string
//   "event_key"      — i1 = listener slot, i2 = event index    -> int as string
//   "event_value"    — i1 = listener slot, i2 = event index    -> int as string
//   "event_reason"   — i1 = listener slot, i2 = event index    -> "CAPACITY"|"TTL_EXPIRED"|"EXPLICIT_DELETE"

struct LruOp {
    string kind;
    int i1;
    int i2;
    int i3;
    int i4;
};

struct CapturedEvent { int key; int value; EvictionReason reason; };

class CapturingListener : public EvictionListener {
public:
    vector<CapturedEvent> events;
    void onEviction(int key, int value, EvictionReason r) override {
        events.push_back({key, value, r});
    }
};

static const char* reason_str(EvictionReason r) {
    switch (r) {
        case EvictionReason::CAPACITY:        return "CAPACITY";
        case EvictionReason::TTL_EXPIRED:     return "TTL_EXPIRED";
        case EvictionReason::EXPLICIT_DELETE: return "EXPLICIT_DELETE";
    }
    return "UNKNOWN";
}

vector<string> lru_simulate(vector<LruOp> ops) {
    vector<string> out;
    unique_ptr<LRUCache> cache;
    vector<unique_ptr<CapturingListener>> listeners;
    auto ensure_listener = [&](int idx) {
        while ((int)listeners.size() <= idx)
            listeners.push_back(unique_ptr<CapturingListener>(new CapturingListener()));
    };
    for (const auto& op : ops) {
        const string& k = op.kind;
        if (k == "new") {
            cache.reset(new LRUCache(op.i1));
            listeners.clear();
            out.push_back("ok");
        } else if (k == "put") {
            cache->put(op.i1, op.i2);
            out.push_back("ok");
        } else if (k == "put_t") {
            cache->put(op.i1, op.i2, (long)op.i3, op.i4);
            out.push_back("ok");
        } else if (k == "get") {
            out.push_back(to_string(cache->get(op.i1)));
        } else if (k == "get_t") {
            out.push_back(to_string(cache->get(op.i1, (long)op.i2)));
        } else if (k == "delete") {
            out.push_back(cache->deleteKey(op.i1) ? "ok" : "fail");
        } else if (k == "size") {
            out.push_back(to_string(cache->size()));
        } else if (k == "add_listener") {
            ensure_listener(op.i1);
            cache->addEvictionListener(listeners[op.i1].get());
            out.push_back("ok");
        } else if (k == "remove_listener") {
            if (op.i1 < (int)listeners.size())
                cache->removeEvictionListener(listeners[op.i1].get());
            out.push_back("ok");
        } else if (k == "events_count") {
            out.push_back(op.i1 < (int)listeners.size()
                          ? to_string((int)listeners[op.i1]->events.size())
                          : "0");
        } else if (k == "event_key") {
            if (op.i1 < (int)listeners.size() && op.i2 < (int)listeners[op.i1]->events.size())
                out.push_back(to_string(listeners[op.i1]->events[op.i2].key));
            else out.push_back("");
        } else if (k == "event_value") {
            if (op.i1 < (int)listeners.size() && op.i2 < (int)listeners[op.i1]->events.size())
                out.push_back(to_string(listeners[op.i1]->events[op.i2].value));
            else out.push_back("");
        } else if (k == "event_reason") {
            if (op.i1 < (int)listeners.size() && op.i2 < (int)listeners[op.i1]->events.size())
                out.push_back(reason_str(listeners[op.i1]->events[op.i2].reason));
            else out.push_back("");
        } else {
            out.push_back("unknown:" + k);
        }
    }
    return out;
}

// ─── Main ────────────────────────────────────────────────────────────────────

#ifndef RUNNING_TESTS
int main() {
    LRUCache cache(3);

    cache.put(1, 10);
    cache.put(2, 20);
    cache.put(3, 30);
    cout << "get(1) = " << cache.get(1) << endl;  // 10
    cache.put(4, 40);  // evicts 2
    cout << "get(2) = " << cache.get(2) << endl;  // -1 (evicted)
    cout << "get(3) = " << cache.get(3) << endl;  // 30
    cout << "get(4) = " << cache.get(4) << endl;  // 40

    return 0;
}
#endif

File: problems/tier2-intermediate/017-lru-cache/solution.java
// LRU Cache — Solution (Java)
import java.util.*;

class LruOp {
    public String kind;
    public int i1, i2, i3, i4;

    public LruOp(String kind, int i1, int i2, int i3, int i4) {
        this.kind = kind; this.i1 = i1; this.i2 = i2; this.i3 = i3; this.i4 = i4;
    }
}

enum EvictionReason { CAPACITY, TTL_EXPIRED, EXPLICIT_DELETE }

class CapturedEvent {
    public int key;
    public int value;
    public EvictionReason reason;
    public CapturedEvent(int k, int v, EvictionReason r) { key = k; value = v; reason = r; }
}

interface EvictionListener {
    void onEviction(int key, int value, EvictionReason reason);
}

class LruNode {
    int key, value;
    long expiresAt;
    LruNode prev, next;
    LruNode(int k, int v, long exp) { key = k; value = v; expiresAt = exp; }
}

class LRUCache {
    int capacity;
    Map<Integer, LruNode> cache = new HashMap<>();
    LruNode head, tail;
    List<EvictionListener> listeners = new ArrayList<>();

    public LRUCache(int cap) {
        this.capacity = cap;
        head = new LruNode(0, 0, 0);
        tail = new LruNode(0, 0, 0);
        head.next = tail;
        tail.prev = head;
    }

    private void addToFront(LruNode node) {
        node.next = head.next;
        node.prev = head;
        head.next.prev = node;
        head.next = node;
    }

    private void removeNode(LruNode node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void moveToFront(LruNode node) { removeNode(node); addToFront(node); }

    private boolean isExpired(LruNode node, long currentTime) {
        return node.expiresAt > 0 && currentTime >= node.expiresAt;
    }

    private void notifyListeners(int key, int value, EvictionReason reason) {
        // Snapshot to avoid concurrent modification
        List<EvictionListener> copy = new ArrayList<>(listeners);
        for (EvictionListener l : copy) l.onEviction(key, value, reason);
    }

    private void evictNode(LruNode node, EvictionReason reason) {
        removeNode(node);
        cache.remove(node.key);
        notifyListeners(node.key, node.value, reason);
    }

    private void evictLRU() {
        if (head.next == tail) return;
        LruNode lru = tail.prev;
        evictNode(lru, EvictionReason.CAPACITY);
    }

    public int get(int key) {
        LruNode node = cache.get(key);
        if (node == null) return -1;
        moveToFront(node);
        return node.value;
    }

    public void put(int key, int value) {
        LruNode existing = cache.get(key);
        if (existing != null) {
            existing.value = value;
            moveToFront(existing);
            return;
        }
        if (cache.size() >= capacity) evictLRU();
        LruNode newNode = new LruNode(key, value, 0);
        addToFront(newNode);
        cache.put(key, newNode);
    }

    public int getT(int key, long currentTime) {
        LruNode node = cache.get(key);
        if (node == null) return -1;
        if (isExpired(node, currentTime)) {
            evictNode(node, EvictionReason.TTL_EXPIRED);
            return -1;
        }
        moveToFront(node);
        return node.value;
    }

    public void putT(int key, int value, long currentTime, int ttl) {
        long expiresAt = (ttl > 0) ? currentTime + ttl : 0;
        LruNode existing = cache.get(key);
        if (existing != null) {
            if (isExpired(existing, currentTime)) {
                evictNode(existing, EvictionReason.TTL_EXPIRED);
            } else {
                existing.value = value;
                existing.expiresAt = expiresAt;
                moveToFront(existing);
                return;
            }
        }
        if (cache.size() >= capacity) evictLRU();
        LruNode newNode = new LruNode(key, value, expiresAt);
        addToFront(newNode);
        cache.put(key, newNode);
    }

    public boolean deleteKey(int key) {
        LruNode node = cache.get(key);
        if (node == null) return false;
        evictNode(node, EvictionReason.EXPLICIT_DELETE);
        return true;
    }

    public int size() { return cache.size(); }

    public void addEvictionListener(EvictionListener listener) { listeners.add(listener); }

    public void removeEvictionListener(EvictionListener listener) {
        listeners.remove(listener);
    }
}

class CapturingListener implements EvictionListener {
    public List<CapturedEvent> events = new ArrayList<>();
    @Override
    public void onEviction(int key, int value, EvictionReason reason) {
        events.add(new CapturedEvent(key, value, reason));
    }
}

public class Solution {
    public static List<String> lru_simulate(List<LruOp> ops) {
        List<String> out = new ArrayList<>();
        LRUCache cache = null;
        List<CapturingListener> listeners = new ArrayList<>();

        for (LruOp op : ops) {
            String k = op.kind;
            if ("new".equals(k)) {
                cache = new LRUCache(op.i1);
                listeners = new ArrayList<>();
                out.add("ok");
            } else if ("put".equals(k)) {
                cache.put(op.i1, op.i2);
                out.add("ok");
            } else if ("put_t".equals(k)) {
                cache.putT(op.i1, op.i2, (long) op.i3, op.i4);
                out.add("ok");
            } else if ("get".equals(k)) {
                out.add(Integer.toString(cache.get(op.i1)));
            } else if ("get_t".equals(k)) {
                out.add(Integer.toString(cache.getT(op.i1, (long) op.i2)));
            } else if ("delete".equals(k)) {
                out.add(cache.deleteKey(op.i1) ? "ok" : "fail");
            } else if ("size".equals(k)) {
                out.add(Integer.toString(cache.size()));
            } else if ("add_listener".equals(k)) {
                while (listeners.size() <= op.i1) listeners.add(new CapturingListener());
                cache.addEvictionListener(listeners.get(op.i1));
                out.add("ok");
            } else if ("remove_listener".equals(k)) {
                if (op.i1 < listeners.size()) cache.removeEvictionListener(listeners.get(op.i1));
                out.add("ok");
            } else if ("events_count".equals(k)) {
                out.add(op.i1 < listeners.size() ? Integer.toString(listeners.get(op.i1).events.size()) : "0");
            } else if ("event_key".equals(k)) {
                if (op.i1 < listeners.size() && op.i2 < listeners.get(op.i1).events.size())
                    out.add(Integer.toString(listeners.get(op.i1).events.get(op.i2).key));
                else out.add("");
            } else if ("event_value".equals(k)) {
                if (op.i1 < listeners.size() && op.i2 < listeners.get(op.i1).events.size())
                    out.add(Integer.toString(listeners.get(op.i1).events.get(op.i2).value));
                else out.add("");
            } else if ("event_reason".equals(k)) {
                if (op.i1 < listeners.size() && op.i2 < listeners.get(op.i1).events.size())
                    out.add(listeners.get(op.i1).events.get(op.i2).reason.name());
                else out.add("");
            } else {
                out.add("unknown:" + k);
            }
        }
        return out;
    }
}
